Periodically-called externs

isaac-ped opened the pull request:
This adds functionality to bm_sim and the simple_switch target to register an function to be called periodically with a given interval.

A function declared outside of a primitive will always be called in that interval:

```cpp
static void run_always() {
    BMLOG_DEBUG("This function will run once a second");
}

REGISTER_PERIODIC_ACTION(run_always, std::chrono::seconds(1));
```

A function declared inside of a primitive will only be called if that primitive is used as an extern somewhere in the compiled P4 file:

```cpp
class timeout_extern : public ActionPrimitive<> {

    void operator() () {
        BMLOG_DEBUG("Calling timeout_extern");
    }


    static void periodic() {
        BMLOG_DEBUG("This will be called 4 times a second if timeout_extern() is used in P4");
    }

    REGISTER_PERIODIC_ACTION(periodic, std::chrono::milliseconds(250));
};
```

antoninbas commented on include/bm/bm_sim/actions.h:
```diff
@@ -96,6 +97,42 @@ class NamedCalculation;
+
+ private:
+  unsigned int tick;
+  std::vector<PeriodicAction> list_{};
```
Wouldn't a priority queue be more appropriate for this? The actions can be ordered based on the next execution timestamp (last + period) and you can sleep until the next action (first element of the queue) is ready to be executed. You wouldn't need a min interval and you wouldn't need to iterate through the whole list of actions. I think you would get O(logN) complexity every time an action is executed.

isaac-ped replied:
Absolutely true. Will update to use priority queue shortly.

antoninbas commented on include/bm/bm_sim/actions.h:
```diff
@@ -96,6 +97,42 @@ class NamedCalculation;
+};
+
+// Contains the list of the periodically executed actions
+class PeriodicActionsList {
```
This class is not thread safe. Have you given some thought to what happens in case of P4 program swap?

isaac-ped replied:
I'm unsure how the program swap occurs, and why it would cause concurrent access of this structure. Could you elaborate a bit on what causes this class to be less safe than ActionsOpcodesMap, or point me to documentation on the program swap procedure?

antoninbas replied:
Swapping the P4 program is done by loading a new JSON file (generated by the compiler). Unfortunately I do not believe there is any centralized documentation fore this, only comments scattered in the switch.h source code. It is very similar to the P4Runtime ForwardingPipelineConfig swap if you're familiar with P4Runtime. It is a 2 step operation:
* first, `load_new_config` is called (https://github.com/p4lang/behavioral-model/blob/master/include/bm/bm_sim/switch.h#L789) to load a new JSON (sent through RPC). All packet processing still uses the old JSON, but all subsequent runtime operations (e.g. table entry adds) are done on the new JSON
* when the application is ready, it calls `swap_configs` (https://github.com/p4lang/behavioral-model/blob/master/include/bm/bm_sim/switch.h#L792) which lets bmv2 know that the new JSON & associated runtime state are ready to be used for packet processing
There is actually a 3rd step. `swap_configs` only signals to bmv2 that configurations should be swapped. But the target (e.g. simple_switch) has final say and controls exactly when the swap occurs. This is because the target may have to update some internal state, e.g. pointers that may be invalidated by the swap. This is detailed here: https://github.com/p4lang/behavioral-model/blob/master/include/bm/bm_sim/switch.h#L49

Now back to your PR. `REGISTER_PERIODIC_ACTION` is actually executed when the first instance of a given primitive (extern function) is encountered while loading the JSON configuration. You have a single global instance of `PeriodicActionsList` with a thread periodically executing `next()`. What if `next()` is called while the new config is loaded and `REGISTER_PERIODIC_ACTION`? So that's why you need to make `PeriodicActionsList` thread safe.

When a new configuration is enabled, you need to cancel all the tasks for the previous P4 and start tasks for the new P4 IMO (the 2 programs may not be using the same extern functions). The only way to make that work is to tie the lifetime of `PeriodicActionsList` with that of `P4Objects`, by letting the `P4Objects` instance own the `PeriodicActionsList` instance just like it owns the `LearnEngine` instance. When configs are swapped, the `PeriodicActionsList` instances will be swapped as well. Now there is one issue with this approach: what about the periodic tasks registered outside of primitive / extern objects?

That leads us to what I think is the right approach. Let a `PeriodicActionsList` instance manage its own thread (like `LearnEngine`). Keep using the singleton pattern; even though it would make more sense to give ownership to `SwitchWContexts` IMO, it may be unpractical to achieve. But modify the `REGISTER_PERIODIC_ACTION` mechanics. Maybe it could look a lot like starting a thread actually.
```
// should not be copyable
// this is merely a skeleton
struct PeriodicTask {
    PeriodicTask(...) { PeriodicActionsList::get_instance()->register(...); }
    ~PeriodicTask() { cancel(); }
    void cancel() { PeriodicActionsList::get_instance()->unregister(...); }
};

class timeout_extern : public ActionPrimitive<> {
    timeout_extern() {
        // the task will be unregistered when the object is destroyed
        task = PeriodicTask(periodic, interval, this);
     }

    void operator() () {
        BMLOG_DEBUG("Calling timeout_extern");
    }

    // it is not a good idea to make this static, as it is not a good idea to have any shared static state in primitive instances
    // static void periodic() {
    void periodic() {
        BMLOG_DEBUG("This will be called 4 times a second if timeout_extern() is used in P4");
    }

    PeriodicTask task;
};
```
For a global static task, you could do `new PeriodicTask(...)` and never destroy it, or you could give ownership to the target class (e.g. `SimpleSwitch`).

This is still more convenient and efficient than having to manage a thread yourself for every periodic task you want to execute. You will also not have to worry about P4 program swap, as its `P4Objects` instance should have its own instance of each primitive in use.

antoninbas replied:
If you do agree to go this route, I believe you should put the new classes in their own source files, not in actions.h / actions.cpp

antoninbas commented on targets/simple_switch/simple_switch.cpp:
```diff
@@ -188,6 +188,7 @@ SimpleSwitch::start_and_return_() {
     threads_.push_back(std::thread(&SimpleSwitch::egress_thread, this, i));
   }
   threads_.push_back(std::thread(&SimpleSwitch::transmit_thread, this));
+  threads_.push_back(std::thread(&SimpleSwitch::periodic_thread, this));
```
I'm not sure why we need any change in simple_switch at all. Shouldn't the periodic thread be managed entirely by the `SwitchWContexts` base class?

isaac-ped replied:
Documentation for in `SwitchWContexts` states that the `start_and_return()` function of an individual target should start all processing threads. Is it acceptable to do processing in `SwitchWContexts` as well?

antoninbas replied:
There are many threads that are already managed by `SwitchWContexts`. All targets can benefit from this kind of periodic task registration and you can hide thread management from the target (it's already done for learning for example).

antoninbas commented on targets/simple_switch/simple_switch.cpp:
```diff
@@ -267,6 +269,16 @@ SimpleSwitch::get_time_since_epoch_us() const {
+SimpleSwitch::periodic_thread() {
+  auto actions = bm::PeriodicActionsList::get_instance();
+  while (!exiting) {
+    if (actions->next() != true) {
```
It seems that if there is no periodic action registered, the thread exits right away. But what if I load a new P4 program that uses an extern with a periodic action?

isaac-ped commented:
Thank you for all of the detailed comments. I believe I have fixed the thread-safety problems, and the periodic tasks should be registered automatically as the externs they are tied to are loaded and unloaded by new configs.

Similar to before, 

```cpp
void tick() {
    std::cout << "Runs once a second" << std::endl;
}

REGISTER_GLOBAL_PERIODIC(tick, std::chrono::seconds(1));

class test_extern : public ActionPrimitive<>{
    void periodic() {
        std::cout << "Runs 4 times a second, if test_extern() is used in P4" << std::endl;
    }

    REGISTER_EXTERN_PERIODIC(test_extern, periodic, std::chrono::milliseconds(250));
}
```

I have also changed the underlying structure to be a priority_queue as you suggested.

I am still having trouble, however, figuring out where in SwitchWContexts to start the thread. All that needs to be added is a `PeriodicTaskList::get_instance()->start()` somewhere.
I could also make the PeriodicTaskList start its thread automatically upon creation.

antoninbas commented on include/bm/bm_sim/periodic_task.h:
```diff
@@ -0,0 +1,92 @@
+  ~PeriodicTask();
+};
+
+// This smells a bit (binding 'this' in inline member initialization),
```
This exactly why I suggested doing the registration in the constructor, I'm not sure why you are insisting on keeping macros.

isaac-ped replied:
Did not realize that you were suggesting removing the macros altogether. Will remove macros in favor of the registration method you suggested above.

antoninbas commented on include/bm/bm_sim/periodic_task.h:
```diff
@@ -0,0 +1,92 @@
+
+// This smells a bit (binding 'this' in inline member initialization),
+// but it compiles and I haven't found documentation against it
+#define REGISTER_EXTERN_PERIODIC(extrn_name, fn, interval) \
```
Even though `extern` is a reserved work in C, is there a problem with using `extern_name`?

antoninbas commented on include/bm/bm_sim/periodic_task.h:
```diff
@@ -0,0 +1,92 @@
+                   interval);
+
+#define REGISTER_GLOBAL_PERIODIC(fn, interval) \
+  static PeriodicTask fn## __task = PeriodicTask(#fn, fn, interval);
```
This is not great (and again why I suggested not using these macros). Even though you may be able to guarantee that the `PeriodicTaskList` instance will be destroyed after all the `PeriodicTask` objects (because you may be able to guarantee that the `PeriodicTaskList` instance is constructed first), can't you just avoid static global variables?

antoninbas commented on include/bm/bm_sim/periodic_task.h:
```diff
@@ -0,0 +1,92 @@
+  static PeriodicTask fn## __task = PeriodicTask(#fn, fn, interval);
+
+class PeriodicTaskList {
+  std::mutex queue_mutex;
```
can you order you data members and function members according to the Google style guide?

antoninbas commented on include/bm/bm_sim/periodic_task.h:
```diff
@@ -0,0 +1,92 @@
+
+  TaskQueue task_queue;
+
+  std::atomic<bool> running;
```
Not sure why you need 2 variables (`running` and `exiting`)
Also since you have a mutex anyway, I don't think there is a big advantage in making them atomic rather than just using the mutex.

antoninbas commented on src/bm_sim/periodic_task.cpp:
```diff
@@ -0,0 +1,170 @@
+PeriodicTaskList::contains_task(PeriodicTask *task) {
+  std::lock_guard<std::mutex> lock(queue_mutex);
+
+  TaskQueue tmp;
```
This is why it probably makes sense to differentiate between a *registered task* and the *next execution of a task*
You could create a new `struct` for *next execution of task* and a new data structure mapping task pointer to *next execution of task* pointer. This would let you check in constant time if a task exists and would also let you do constant-time "lazy" removal of a task.
However, performance is not really a concern for task registration & removal so I won't block the PR for this.

antoninbas commented on src/bm_sim/periodic_task.cpp:
```diff
@@ -0,0 +1,170 @@
+
+void
+PeriodicTaskList::start() {
+  if (running) {
```
needs to be protected by locking the mutex

antoninbas commented on src/bm_sim/periodic_task.cpp:
```diff
@@ -0,0 +1,170 @@
+    if (!task_queue.empty()) {
+      next = task_queue.top()->next;
+    } else {
+      next = std::chrono::system_clock::now() + kDefaultTimeout;
```
sorry if I'm missing something but why do we need a default timeout, given that we use a condition variable?

antoninbas commented on src/bm_sim/switch.cpp:
```diff
@@ -72,6 +73,8 @@ SwitchWContexts::receive(port_t port_num, const char *buffer, int len) {
 void
 SwitchWContexts::start_and_return() {
+  // Starts any registered periodically-executing externs
+  PeriodicTaskList::get_instance().start();
```
Ir would probably beet to put this with the `start` call underneath, unless you have a good reason to start execution before a P4 program is loaded?

antoninbas commented on tests/test_periodic_tasks.cpp:
```diff
@@ -0,0 +1,108 @@
+  PeriodicIncrementor incrementor(kInterval, &i);
+  std::this_thread::sleep_for(kSleepTime);
+
+  ASSERT_EQ(i, 0);
```
I didn't know that when I started writing GTests, but unless a violation of the expectation causes the rest to crash, I believe `EXPECT_*` macros are preferred.

isaac-ped commented on include/bm/bm_sim/periodic_task.h:
```diff
@@ -0,0 +1,120 @@
+//!   PeriodicTask task;
+//! }
+//! @endcode
+class PeriodicTask {
```
Makes sense. I will remove the constructors.

isaac-ped commented on include/bm/bm_sim/periodic_task.h:
```diff
@@ -0,0 +1,120 @@
+
+//! Initializes and registers a task that is to be executed periodically
+//! at a fixed interval. Task will execute during the lifetime of the object.
+//! Make a member of an ActionPrimitive class to tie the execution to the
```
Was trying to explain how making an PeriodicTask a member of an ActionPrimitive object will ensure that the task is only executed if the extern represented by the ActionPrimitive is used in P4. Will clarify in the comment.

isaac-ped commented on src/bm_sim/periodic_task.cpp:
```diff
@@ -0,0 +1,170 @@
+    return false;
+  }
+
+  std::lock_guard<std::mutex> lock(queue_mutex);
```
Right you are! Missed that. Will fix.

isaac-ped commented on include/bm/bm_sim/periodic_task.h:
```diff
@@ -0,0 +1,120 @@
+/* Copyright 2013-present Barefoot Networks, Inc.
```
I am waiting to hear back about using my organizations name in the copyright. Will you need a CLA to accept the changes, and will it be worded as in: https://p4.org/assets/P4_Language_Consortium_Membership_Agreement.pdf ?

antoninbas replied:
If your organization is not a p4.org member, you are making this contribution as an individual and you should use your name in the copyright.
If your organization is not a member(which is usually the most convenient way to contribute) and not interested in becoming one, you can sign the CLA as an individual. I think you can fill / sign the PDF and send it to membership@p4.org.

isaac-ped commented:
My apologies for the long delay.
I am still waiting for a decision regarding the copyright. For the moment, I have fixed the other issues.

I understand your comments regarding the different data structures for a registered task vs. task execution. Because (as you said) task registration performance is not a concern, I did not make that update at this time.

antoninbas commented on include/bm/bm_sim/periodic_task.h:
```diff
@@ -0,0 +1,127 @@
+class PeriodicTask {
+ public:
+  // Friend so that PeriodicTaskList can call execute()
+  friend class PeriodicTaskList;
```
Not an issue

antoninbas commented on include/bm/bm_sim/periodic_task.h:
```diff
@@ -0,0 +1,127 @@
+
+  // Deleting copy-constructor and copy-assignment
+  PeriodicTask(const PeriodicTask&) = delete;
+  PeriodicTask& operator= (const PeriodicTask&) = delete;
```
you actually must make your class non-movable as well

antoninbas commented on include/bm/bm_sim/periodic_task.h:
```diff
@@ -0,0 +1,127 @@
+  std::thread periodic_thread;
+  bool running;
+  std::mutex queue_mutex;
+  std::condition_variable cv;
```
I recommend making `mutex` and `cv` mutable for future-proofness

isaac-ped replied:
I have taken your suggestion, but am not sure what exactly you mean by future-proofness.
Do you mind clarifying a little or providing a link on the topic, if only for my edification? I have read up on the interactions between const functions and mutable members, but am unsure how it applies.
By this, do you mean that the public functions of the PeriodicTaskList should have been `const`, and as such the mutex and cv should be mutable? Or is it simply because of the guarantee that `mutex` and `cv` are already thread safe, and thus do not need the protection that `const` offers?

antoninbas commented on src/bm_sim/periodic_task.cpp:
```diff
@@ -0,0 +1,165 @@
+
+  // Do not add a task twice
+  if (contains) {
+    BMLOG_DEBUG("Warning: Task {} already exists in periodic task queue",
```
this deserves to be a warning IMO

antoninbas commented on src/bm_sim/periodic_task.cpp:
```diff
@@ -0,0 +1,165 @@
+
+  cv.notify_all();
+  if (!removed) {
+    BMLOG_DEBUG("Warning: Task {} does not exist in periodic task queue",
```
this deserves to be a warning as well

antoninbas commented on src/bm_sim/periodic_task.cpp:
```diff
@@ -0,0 +1,165 @@
+PeriodicTaskList::start() {
+  std::lock_guard<std::mutex> lock(queue_mutex);
+  if (running) {
+    BMLOG_DEBUG("Warning: Tried to start already-running periodic task loop");
```
make it a warning and remove "Warning" from the message string...

antoninbas commented on src/bm_sim/periodic_task.cpp:
```diff
@@ -0,0 +1,165 @@
+  {
+    std::lock_guard<std::mutex> lock(queue_mutex);
+    if (!running) {
+      BMLOG_DEBUG("Warning: Tried to join non-running periodic thread");
```
make it a warning and remove "Warning" from the message string...

isaac-ped commented:
Sorry for all the delays. Still waiting to hear from higher-ups about copyright.

isaac-ped commented:
Still waiting to hear about copyright (I had never expected the answer to take this long). Will (hopefully) hear back in the next couple of days. Thanks for your patience, and for all of your help and explanation!

isaac-ped commented:
I have updated the copyright messages and merged the current master branch into my fork. Do let me know if any other updates are needed.

isaac-ped commented:
I am so sorry for the delay. The membership agreement had not been executed as of my last message, but has now been completed.
